### PR TITLE
Add function parsing to chat request parser and abstract chat agent

### PR DESCRIPTION
**Pull request by lucas-koehler**
#### What it does

Implements https://github.com/eclipsesource/osweek-2024/issues/101

- Parse function/tools lead by a `~`. E.g. `~getWorkspaceFileList`
- Put parsed tool requests in the ParsedChatRequest to facilitate reading them in agents.
- Exten abstract chat agent to hand over parsed tools to the language model.
- Improve tool request parameters type
- Add autocompletion for tools/functions to the chat input

#### How to test

- Open a chat
- Type `~` to see tool completion. select one
- prompt related to the tool (e.g. describe workspace content)

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/ai-chat/src/common/chat-request-parser.ts [hunk 1]
```diff
@@ -181,4 +196,21 @@ export class ChatRequestParserImpl {
...
         return new ChatRequestVariablePart(varRange, name, variableArg);
     }
+
+    private tryParseFunction(message: string, offset: number): ChatRequestFunctionPart | undefined {
+        const nextFunctionMatch = message.match(functionReg);
+        if (!nextFunctionMatch) {
+            return;
+        }
+
+        const [full, name] = nextFunctionMatch;
```
Comment (eneufeld) [inline packages/ai-chat/src/common/chat-request-parser.ts:206]: name or id?

**Review by eneufeld — COMMENTED** [review]
Looks good only a nitpick about naming

**Comment by lucas-koehler** [comment]
@eneufeld I fixed the naming to reflect that the id is used.

**Review by sdirix — APPROVED** [review]
Works for me

**Pull request merged**